### PR TITLE
Triton norms dispatch refactor

### PR DESCRIPTION
**Pull request by Micky774**
# Description

This PR disentangles the backend triton implementation from the front-end API, creating a unified intermediate `te_norm_fwd_triton` which is a generalized dispatch function. This PR is fully backwards compatible, as `te_rmsnorm_fwd_triton` and `te_layernorm_fwd_triton` are preserved and implemented as thin wrappers around `te_norm_fwd_triton`.

This way, when bugs appear, we fix them **once** without needing to duplicate across norms.

Consequently, there are some changes to the imports to accommodate this restructuring. This PR also includes a minor cleanup/simplification of previously redundant behavior in the layernorm fwd implementation, as well as support for `Float8CurrentScalingQuantizer`.

FWIW I don't think we can apply a similar unification to the backwards passes, as it seems that -- at least for layernorm -- the backwards implementations are pretty specialized and have asymmetric heuristics.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by Micky774** [comment]
Note that currently some of the layernorm tests are failing, but they're citing `NaN` vals in the `expected` tensor, i.e. the HIP reference kernel. I tried including @alextmagro's PR #303 but it still fails. @alextmagro have you seen such an error as well? Is it something related?

**Comment by alextmagro** [comment]
> Note that currently some of the layernorm tests are failing, but they're citing `NaN` vals in the `expected` tensor, i.e. the HIP reference kernel. I tried including @alextmagro's PR #303 but it still fails. @alextmagro have you seen such an error as well? Is it something related?

I haven't seen anything like that.

File: transformer_engine/pytorch/triton_kernels/norms_common.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
+import torch
```
Comment (ipanfilo) [inline transformer_engine/pytorch/triton_kernels/norms_common.py:4]: Copyright

File: tests/pytorch/triton_kernels/test_norms.py [hunk 2]
```diff
@@ -17,13 +17,11 @@
 )
 from transformer_engine.pytorch.tensor.float8_tensor import Float8Quantizer, Float8Tensor
 from transformer_engine.pytorch.tensor.mxfp8_tensor import MXFP8Quantizer, MXFP8Tensor
-from transformer_engine.pytorch.triton_kernels.rmsnorm import (
```
Comment (ipanfilo) [inline tests/pytorch/triton_kernels/test_norms.py:20]: So layernorm and rmnsorm now contains kernels only and norms.py bears launchers?

Comment (Micky774) [reply tests/pytorch/triton_kernels/test_norms.py:20]: Yes. It's mainly for the `fwd` pass functions, but I moved over the `bwd` pass ones too for consistency. I don't have high hopes that we can merge the `bwd` pass functions into a single dispatch since they are pretty meaningfully distinct.

**Comment by Micky774** [comment]
Turns out _something_ in this PR makes it so that the layernorm kernel has bad memory behavior. Specifically, it mutates either the weight tensor, or the bias tensor in the test. This happens, I believe, because they are allocated on GPU contiguously wrt each other (i.e. first input array, then gamma, then bias) which leads me to suspect that there's some kind of masking problem with the layernorm kernel, but I have not been able to pinpoint it yet.

Everything seems to work on `dev`, but I don't have any functional changes aside from a logical simplification of non-atomic layernorm fwd cases. Most of it is just variable renaming.

**Comment by wenchenvincent** [comment]
@Micky774 Could you remind me of what we had decided on this PR?

**Comment by Micky774** [comment]
> @Micky774 Could you remind me of what we had decided on this PR?

@wenchenvincent last we talked about this, I think @matthiasdiener was supposed to eventually take it over. There's currently a bug exposed by this PR that will require a bit of work to resolve I think. There's some kind of memory mismanagement occurring, where output tensors' memory is being overwritten after being produced.

**Comment by Micky774** [comment]
~~Note that I skip the tests where the HIP kernel generates `nan` values, but if desired, I can instead test against a reference torch implementation. I prefer the skip, since it highlights the fact that the test isn't fully-enabled as opposed to us forgetting that certain cases were handled with a torch reference implementation...~~

This was a manifestation of the aforementioned bug

**Comment by Micky774** [comment]
cc: @wenchenvincent @wangye805 
I finally found the underlying issue. There was a problem with the sizing of the `amax` array in the layernorm kernel, which led to unsafe memory writes corrupting adjacent data. I've corrected this.

The PR is ready for review!

**Comment by wenchenvincent** [comment]
LGTM. @ipanfilo You reviewed it a while ago. Do you have further comments?

**Review by aris134 — APPROVED** [review]

File: tests/pytorch/triton_kernels/test_norms.py [hunk 3]
```diff
@@ -1,29 +1,30 @@
-# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
+# Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_norms.py]: 2025-2026

File: tests/pytorch/triton_kernels/test_norms.py [hunk 4]
```diff
@@ -1,29 +1,30 @@
...
 # License for AMD contributions = MIT. See LICENSE for more information
 
 
+import math
 import os
 import torch
 import pytest
 from functools import partial
 from itertools import product
+from torch.utils.cpp_extension import IS_HIP_EXTENSION
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_norms.py]: This entire file is ROCM specific. Basically we can assume IS_HIP_EXTENSION is true when running this pytest

File: tests/pytorch/triton_kernels/test_norms.py [hunk 5]
```diff
@@ -437,67 +438,66 @@ def _compare_output_tensors(
...
                 expected=out_hip._scale_inv,
                 msg=lambda msg: f"Output scale inverse does not match triton <-> hip\n\n{msg}\n",
             )
         elif quantization == "mxfp8":
             has_rscale_triton = out_triton._rowwise_scale_inv is not None
             has_rscale_hip = out_hip._rowwise_scale_inv is not None
+
+            # The scale_inv values may differ slightly, but will still dequantize close enough to 
+            # pass the earlier comparisons.
+            compare_func = partial(te_compare_results, atol=1, rtol=0, use_torch_semantics=True)
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_norms.py:459]: For mxfp8 data and scale inv comparison, we can reuse the same logic in cpp gtest: https://github.com/ROCm/TransformerEngine/blob/0dfee560c7ad795fe554ebde7e0c9d9d30116d2e/tests/cpp/test_common.cu#L730
https://github.com/ROCm/TransformerEngine/blob/0dfee560c7ad795fe554ebde7e0c9d9d30116d2e/tests/cpp/operator/test_cast_mxfp8.cu#L331-L355

File: tests/pytorch/triton_kernels/test_norms.py [hunk 6]
```diff
@@ -437,67 +438,66 @@ def _compare_output_tensors(
...
+
+            # The MXFP8 tensors carry their scale_inv values in a padded
+            # format, hence we must omit the padded values.
+            input_shape = out_triton.shape
+            unpadded_scale_inv_shape = (math.prod(input_shape[:-1]), input_shape[-1] // MXFP8_BLOCK_SCALING_SIZE)
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_norms.py]: Should we have different shape for row-wise and col-wise scaling?

File: tests/pytorch/triton_kernels/test_utils.py [hunk 7]
```diff
@@ -1,11 +1,11 @@
-# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
+# Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_utils.py]: Same here

File: transformer_engine/pytorch/ops/basic/layer_norm.py [hunk 8]
```diff
@@ -17,7 +17,9 @@
 from transformer_engine_torch import layernorm_bwd, layernorm_fwd
 from torch.utils.cpp_extension import IS_HIP_EXTENSION
 if IS_HIP_EXTENSION:
-    from ...triton_kernels.layernorm import te_layernorm_fwd_triton, te_layernorm_bwd_triton
+    from ...triton_kernels.norms_common import te_layernorm_fwd_triton, te_layernorm_bwd_triton
+from ...fp8 import FP8GlobalStateManager
```
Comment (wangye805) [inline transformer_engine/pytorch/ops/basic/layer_norm.py]: Do we need to import FP8GlobalStateManager and QuantizedTensor here? For both NV upstream and us

**Review by wangye805 — CHANGES_REQUESTED** [review]

File: tests/pytorch/triton_kernels/test_norms.py (hunk 5 above)
Comment (Micky774) [reply tests/pytorch/triton_kernels/test_norms.py:459]: We essentially already do this implicitly by relying on the dequantization of the `MXFP8Tensor`s before comparison. While we could handle this explicitly as in the C tests, I don't think that's necessary given that the dequantization behavior has its own testing which passes. Let me know if you have other thoughts on the matter.

File: transformer_engine/pytorch/ops/basic/layer_norm.py (hunk 8 above)
Comment (Micky774) [reply transformer_engine/pytorch/ops/basic/layer_norm.py]: Removed extra import

File: tests/pytorch/triton_kernels/test_utils.py (hunk 7 above)
Comment (Micky774) [reply tests/pytorch/triton_kernels/test_utils.py]: Fixed

File: tests/pytorch/triton_kernels/test_norms.py (hunk 6 above)
Comment (Micky774) [reply tests/pytorch/triton_kernels/test_norms.py]: We resolve this with re-indexing, but I've updated the variable name for a bit of extra clarity.

File: tests/pytorch/triton_kernels/test_norms.py (hunk 4 above)
Comment (Micky774) [reply tests/pytorch/triton_kernels/test_norms.py]: It was unused anyways -- fixed.

File: tests/pytorch/triton_kernels/test_norms.py (hunk 3 above)
Comment (Micky774) [reply tests/pytorch/triton_kernels/test_norms.py]: Fixed

File: transformer_engine/pytorch/module/_common.py
Comment (ipanfilo) [inline transformer_engine/pytorch/module/_common.py:1]: Copyright date

File: transformer_engine/pytorch/ops/basic/layer_norm.py
Comment (ipanfilo) [inline transformer_engine/pytorch/ops/basic/layer_norm.py:1]: Copyright date

File: transformer_engine/pytorch/ops/basic/layer_norm.py [hunk 9]
```diff
@@ -17,7 +17,8 @@
 from transformer_engine_torch import layernorm_bwd, layernorm_fwd
 from torch.utils.cpp_extension import IS_HIP_EXTENSION
 if IS_HIP_EXTENSION:
-    from ...triton_kernels.layernorm import te_layernorm_fwd_triton, te_layernorm_bwd_triton
+    from ...triton_kernels.norms_common import te_layernorm_fwd_triton, te_layernorm_bwd_triton
+from ...tensor import QuantizedTensor
```
Comment (ipanfilo) [inline transformer_engine/pytorch/ops/basic/layer_norm.py]: Is it needed for this PR?

File: transformer_engine/pytorch/ops/basic/rmsnorm.py
Comment (ipanfilo) [inline transformer_engine/pytorch/ops/basic/rmsnorm.py:1]: Copyright date

File: ci/pytorch.sh [hunk 10]
```diff
@@ -70,9 +70,9 @@ run_test_config(){
     run_default_fa 1 triton_kernels/test_cast.py
     run_default_fa 1 triton_kernels/test_cast_mxfp8.py
     run_default_fa 1 triton_kernels/test_grouped_gemm.py
-    run_default_fa 1 triton_kernels/test_norm_common.py
-    run_default_fa 1 triton_kernels/test_norms.py
-    NVTE_TEST_TRITON_AUTOTUNE=1 run_default_fa_lbl "autotune" 3 triton_kernels/test_norms.py
+    run_default_fa 1 triton_kernels/test_utils.py
+    NVTE_ROCM_ENABLE_MXFP8=1 run_default_fa 1 triton_kernels/test_norms.py
```
Comment (ipanfilo) [inline ci/pytorch.sh:74]: Do those test still cover non-MXFP8 with this env set?

**Review by ipanfilo — CHANGES_REQUESTED** [review]

File: ci/pytorch.sh (hunk 10 above)
Comment (Micky774) [reply ci/pytorch.sh:74]: Yes

File: transformer_engine/pytorch/ops/basic/rmsnorm.py
Comment (Micky774) [reply transformer_engine/pytorch/ops/basic/rmsnorm.py:1]: Done

File: transformer_engine/pytorch/ops/basic/layer_norm.py (hunk 9 above)
Comment (Micky774) [reply transformer_engine/pytorch/ops/basic/layer_norm.py]: No, it was leftover on accident. Removed.

File: transformer_engine/pytorch/ops/basic/layer_norm.py
Comment (Micky774) [reply transformer_engine/pytorch/ops/basic/layer_norm.py:1]: Done

File: transformer_engine/pytorch/module/_common.py
Comment (Micky774) [reply transformer_engine/pytorch/module/_common.py:1]: Done

File: transformer_engine/pytorch/module/layernorm_mlp.py
Comment (ipanfilo) [inline transformer_engine/pytorch/module/layernorm_mlp.py:1]: Copyright date

Comment (Micky774) [reply transformer_engine/pytorch/module/layernorm_mlp.py:1]: Done, along with a few others I missed.

**Review by ipanfilo — APPROVED** [review]

**Review by wangye805 — APPROVED** [review]
LGTM
